src/shared/utils/extractUploads.ts

```diff
@@ -0,0 +1,136 @@
+import { z } from 'zod'
+
+import {
+  UploadErrorCodeEnumSchema,
+  UploadStateEnumSchema,
+  UploadTypeEnumSchema,
+} from 'services/commit'
+import { UploadStateEnum, UploadTypeEnum } from 'shared/utils/commit'
+
+interface ErrorObject {
+  errorCode: z.infer<typeof UploadErrorCodeEnumSchema> | null
+}
+
+export interface Upload {
+  id?: number | null
+  state?: z.infer<typeof UploadStateEnumSchema>
+  provider?: string | null
+  createdAt?: string
+  updatedAt?: string
+  flags?: string[] | null
+  jobCode?: string | null
+  downloadUrl?: string
+  ciUrl?: string | null
+  uploadType?: z.infer<typeof UploadTypeEnumSchema>
+  buildCode?: string | null
+  name?: string | null
+  errors: (ErrorObject | null)[]
+}
+
+function humanReadableOverview(state: string) {
+  if (state === UploadStateEnum.error) return 'errored'
+  if (state === UploadStateEnum.processed) return 'successful'
+  if (state === UploadStateEnum.complete) return 'carried forward'
+  if (state === UploadStateEnum.started) return 'started'
+}
+
+function deleteDuplicateCFFUploads({ uploads }: { uploads: Upload[] }) {
+  const nonCFFlags = new Set()
+
+  // Get all the flags from "uploaded" uploads
+  uploads?.forEach((upload) => {
+    if (
+      upload?.uploadType !== UploadTypeEnum.CARRIED_FORWARD &&
+      upload?.flags
+    ) {
+      upload.flags.forEach((flag) => nonCFFlags.add(flag))
+    }
+  })
+
+  // Filter out uploads that have repeated flags, returning those without duplicates
+  return uploads.filter(
+    (upload) =>
+      !(
+        upload?.uploadType === UploadTypeEnum.CARRIED_FORWARD &&
+        upload?.flags?.some((flag) => nonCFFlags.has(flag))
+      )
+  )
+}
+
+const createUploadGroups = ({ uploads }: { uploads: Upload[] }) => {
```

Comment:
Here we can see the single pass approach, which IMO is just as readable, if not more readable where we have separate "top level" variables for the state counts, provider groups, and errorProviderGroups that all get returned at the the end of the loop